enhancements/sig-architecture/32-extensiblescheduling/32-extensiblescheduling.md

```diff
@@ -0,0 +1,363 @@
+
+There will be a GC() funcion in the framework, being called every 10 minutes. Go through all the Placment CRs and ManagedClusterScalar CRs, and clean not used ManagedClusterScalar CRs.
+
+- Update
+
+ManagedClusterScalar CR's status is updated by customized agent or third party controllers. Out of scope and won't be covered in this proposal.
+
+### Let placement prioritizer support rating clusters with the scalar value of the CR.
+Once ManagedClusterScalar CR is created in PreScore(), a customized agent or third party controllers should watch the change and update the CR status as below.
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: ManagedClusterScalar
+metadata:
+  name: {PrioritizerName}
+  namespace: cluster1
+spec:
+  prioritizerName: {PrioritizerNameInPlacement}
+status:
+  conditions:
+  - lastTransitionTime: "2021-10-28T08:31:39Z"
+    message: ManagedClusterScalar updated successfully
+    reason: ManagedClusterScalarUpdated
+    status: "True"
+    type: ManagedClusterScalarUpdated
+  validUntil: "2021-10-29T18:31:39Z"
+  scalar: 74
+```
+Usually each ManagedClusterScalar CR will be created in more than one cluster namespace, different clusters update the status separately. So the prioritizer plugin which depends on ManagedClusterScalar should follow below logic to get avaliable scalar and sort all the clusters.
+- Prescore
+
+In prescore stage, prioritizer plugin should create the CR before use it (have talked about this in previous section).
+The plugin should also check if 80% of the CRs have a valid scalar (80% is a reasonable threshold hardcode so far, might make it configurable in the future). If yes, the plugin will continue to score the clusters. If not, terminate this scheduling and wait for the next reschedule.
+
+- Score
+
+In score stage, prioritizer plugin get valid scalar from ManagedClusterScalar CR. If no scalar updated in ManagedClusterScalar CR status, treat it as 0.
+After getting all the cluster scalar, the plugin should formalized them from -100 to 100. The maximal will be 100 and minimal is -100, other score distribute in proportion.
+
+- Re-Schedule
+
+Rather than watch ManagedClusterScalar and trigger Score(), that will cause frequently reschedule. The scheduler framework should periodically (eg. every 5 minutes) reschedule the placement.
+
+Future work：there might be a cluster-scoped CRD to configure the threshold of each prioritizer. For example, the frequency to trigger a re-scheudle, the threshold in prescore to treat score as valid.
+
+### Examples
+
+#### An agent update ResourceRatioCPU scalar to API, and placement prioritizer plugin score based on it. (Use Story 1&2)
+1. User create a new placement as below
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: Placement
+metadata:
+  name: placement
+  namespace: ns1
+spec:
+  numberOfClusters: 1
```

Comment:
the `numberOfClusters` should be 2 or bigger, and we will provide sorted clusters, then user can select one (usually the top one).

If the `numberOfClusters` is bigger than 1, then user can do something else with the result, E.g. get the busiest cluster and check it.

Reply:
Do you mean even the desired cluster number is 1 or any other number, the user still wants a list of sort clusters (more than 1) from placement decision, and he picks up one by himself?

Reply:
User will not set `numberOfClusters` for most cases, it is a limitation to the user.

For example, there are total 100 managed clusters, and userA owns 5 of them, what this user want to do is to build a placement with these 5 managed clusters, and sort these clusters by some rules, then select the top one to send workload. This user also need to get the last one in the sorted list to check if there is something wrong with that cluster or do something else.  With `numberOfClusters=1`, the user can only see the status of one cluster, and other clusters will be ignored, this is why I say this is a limitation.

Reply:
As we don't provide a sorted cluster list in the placement decision for now, to handle the above case, a possible solution is to create 2 placements, one sets prioritizer [weight](https://github.com/open-cluster-management-io/api/blob/main/cluster/v1alpha1/types.go#L321) to a positive number and selects the top N, another sets prioritizer weight to a negative number and select the top1. 

Can this satisfy your requirement?